### DISCUSSION TITLE
How do I propagate data to initial state of DataState?

### DISCUSSION THREAD

**Discussion by oyanyev** [nsk90/kstatemachine#59] [Q&A] [answered]
Let say I have hierarchical SM and some data is passed to some inner state. I would like this data propagated to initial state on such inner state.
```
    @Test
    fun propagate_data_to_initial_state() {

        class MoveEvent(override val data: Int) : DataEvent<Int>
        
        val outer = DefaultState("state_initial_1")
        val inner = DefaultDataState("inner_state", defaultData = 0, dataExtractor = defaultDataExtractor())
        val innerInitial = DefaultDataState("inner_initial", defaultData = 0, dataExtractor = defaultDataExtractor())
        
        val sm = createStateMachine {
            addInitialState(outer) {
                initialState("outer_initial")
                addState(inner) {
                    addInitialState(innerInitial)
                }
                dataTransition<MoveEvent, Int> {
                    targetState = inner
                }
            }
        }
        
        sm.processEvent(MoveEvent(99))
        
        assert(inner.data == 99)
        assert(innerInitial.data == 99)
    }
```

**Comment by nsk90** [comment, marked as answer]
Hello, there is not such functionality like Data propagation, sorry. DataState/DataEvent are designed to be type safe API for passing data from event to its target state. Such propagation is out of their scope. 

I suggest simply access `innerInitial`  parent's data via direct reference to `inner` or `parent` property.

```
    "activeStates() not include nested machines states1" {

        class MoveEvent(override val data: Int) : DataEvent<Int>

        val outer = DefaultState("state_initial_1")
        val inner = defaultDataState("inner_state", defaultData = 0) // you can use this helper function
        val innerInitial = DefaultState("inner_initial")

        val sm = createStateMachine {
            addInitialState(outer) {
                initialState("outer_initial")
                addState(inner) {
                    addInitialState(innerInitial) {
                        onEntry {
                            check((parent as DataState<Int>).data == 99)
                        }
                    }
                }
                dataTransition<MoveEvent, Int> {
                    targetState = inner
                }
            }
        }

        sm.processEvent(MoveEvent(99))

        assert(inner.data == 99)
        assert((innerInitial.parent as DataState<Int>).data == 99)
    }
```

If you need more control on how this data should be stored, you should subclass `DefaultState` and mange the data field manually.

**Reply by oyanyev** [reply]
I want to renew the discussion, because i find such propagation necessary.
And I found also that such propagation does exist in form of propagation of TransitionParams:
`setCurrentState(initialState, transitionParams)`
But in DefaultDataState onDoEnter(transitionParams: TransitionParams<*>) a check happens:
```
if (this == transitionParams.direction.targetState) {
...
}
```
which returns false and data is not propagated to initial state.
Please think about changing TransitionParams.direction.targetState to match initial state before 
`setCurrentState(initialState, transitionParams)`
to allow data smoothly propagated to initial state.

Alternatively, DataState.defaultData can be lambda and so can lazily be returned as { parent.data } in DefaultDataState onDoEnter()
```
else { // implicit activation
            _data = lastData
        }
```

lastData initialization
```
    override val lastData: D
        get() = checkNotNull(_lastData ?: defaultData()) {
            "Last data is not available yet in $this, and default data not provided"
        }
```

**Comment by oyanyev** [comment]
Sorry, I see that, in spite state is finished, transitions are triggered from within. Checked that with a test. Let's how it works in a code
